interfaces/builtin/wayland.go

```diff
@@ -19,32 +19,127 @@
 
 package builtin
 
+import (
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
 const waylandSummary = `allows access to compositors supporting wayland protocol`
 
 const waylandBaseDeclarationSlots = `
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection:
+      on-classic: false
 `
 
-const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is a display server and needs to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+# Create the Wayland socket and lock file
+/run/user/[0-9]*/wayland-[0-9]* rw,
+
+# Allow write access to create XDG_RUNTIME_DIR (until lp:1656340 is fixed)
+/run/user/[0-9]*/ w,
```

Comment:
LP: #1656340 was about not creating `/run/user/<uid>/snap.$SNAP`, not `/run/user/<uid>`. The session manager is supposed to create this directory, not the snap. Why did you add this rule?

Reply:
Mir is running as root, and I'm seeing /run/user/0 does not exist.  Well with
        [ -n "$XDG_RUNTIME_DIR" ] && mkdir -p $XDG_RUNTIME_DIR -m 700
I get 
audit[6690]: AVC apparmor="DENIED" operation="mkdir" profile="snap.mir-kiosk.mir-kiosk" name="/run/user/0/" pid=6690 comm="mkdir" requested_mask="c" denied_mask="c" fsuid=0 ouid=0
mir-kiosk.mir-kiosk[6286]: mkdir: cannot create directory ‘/run/user/0’: Permission denied

Session Manager bug?

Reply:
In this case I assume you mean that you are starting mir as a daemon. In this case, there is no user session and IMO it would be snappy's responsibility to create this directory. We had a previous PR for this in snap-confine, but it was never committed for various reasons. @zyga - can you comment? IIRC, the issue was the parent directories of `/run/user/<uid>` but I think if we just create `/run/user/<uid>` as 0700 as per the specification and log (perhaps not fail) if `/run/user` does not exist, we should be fine.

Reply:
Correct assumption. Daemonized Mir just needs an accessible XDG_RUNTIME_DIR somehow.